### DISCUSSION TITLE
Strongswan CHILD_SA established on Openwrt but packets being sent un-encrypted

### DISCUSSION THREAD

**Discussion by abc1233user** [strongswan/strongswan#1647] [Q&A]
I am using strongswan on Openwrt (v22.03) to create a tunnel between two MT-7621 devices. I also have two Ubuntu machines on LAN sides of both these devices.
Here is my network interconnection:

> (192.168.1.100/24) >==Left LAN==< (192.168.1.1/24) ----- (10.10.10.1/24) >===tunnel here===< (10.10.10.2/24) ----- (192.168.2.1/24) >==Right LAN==< (192.168.2.151/24)

Initially I was using the _strongswan-mod-kernel-libipsec_ package and everything was working fine ('ipsec statusall' showed packet increase when pinging from one Ubuntu machine to the other. Also scp and iperf3 worked fine)

Now, I want to use _strongswan-mod-kernel-netlink_ instead of libipsec. CHILD_SA is established in this case, but when I ping from either ubuntu machine to the other, packets are being sent un-encrypted. (I am receiving ping replies from one ubuntu machine to the other, and scp and iperf3 are also working, but packet count on ipsec statusall doesnot increase). If I bring the CHILD_SA down using `ipsec down <name of child_sa>` then the two ubuntu machines cannot reach each other (ping fails)

Here is output of `ipsec statusall` on right side MT-7621 device.  (It is similar on the left one as well)

> Status of IKE charon daemon (strongSwan 5.9.5, Linux 5.10.161, mips):
>   uptime: 50 minutes, since Jan 03 00:49:36 2023
>   worker threads: 10 of 16 idle, 6/0/0/0 working, job queue: 0/0/0/0, scheduled: 2
>   loaded plugins: charon test-vectors pkcs11 aes des rc2 sha2 sha3 sha1 md4 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp curve25519 agent xcbc cmac hmac ctr ccm gcm sqlite attr kernel-netlink resolve socket-default socket-dynamic connmark stroke vici smp updown xauth-generic led addrblock
> Listening IP addresses:
>   10.10.10.2
>   192.168.2.1
>   fd4e:4d87:7d57::1
> Connections:
>          abc:  10.10.10.2...10.10.10.1  IKEv2
>          abc:   local:  [10.10.10.2] uses pre-shared key authentication
>          abc:   remote: [10.10.10.1] uses pre-shared key authentication
>          abc:   child:  192.168.2.0/24 === 192.168.1.0/24 TUNNEL
> Security Associations (1 up, 0 connecting):
>          abc[1]: ESTABLISHED 50 minutes ago, 10.10.10.2[10.10.10.2]...10.10.10.1[10.10.10.1]
>          abc[1]: IKEv2 SPIs: 4e2e86de9515780d_i* 2177b70172de3faa_r, pre-shared key reauthentication in 118 minutes
>          abc[1]: IKE proposal: AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
>          abc{2}:  INSTALLED, TUNNEL, reqid 1, ESP SPIs: c3a47cb6_i cd79a4bf_o
>          abc{2}:  AES_CBC_128/HMAC_SHA1_96/MODP_2048, 0 bytes_i, 0 bytes_o, rekeying in 41 minutes
>          abc{2}:   192.168.2.0/24 === 192.168.1.0/24

`ip xfrm state` on right side MT-7621 device:

> src 10.10.10.2 dst 10.10.10.1
> 	proto esp spi 0xcd79a4bf reqid 1 mode tunnel
> 	replay-window 0 flag af-unspec
> 	auth-trunc hmac(sha1) 0x151ca8d86020a7727f1c881dfdc850168ae81c0c 96
> 	enc cbc(aes) 0x1403ebf8cb288e9f35a8782dc2db4260
> 	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
> src 10.10.10.1 dst 10.10.10.2
> 	proto esp spi 0xc3a47cb6 reqid 1 mode tunnel
> 	replay-window 32 flag af-unspec
> 	auth-trunc hmac(sha1) 0x92d590baae6df8156bde751a0c60bd617e15fdda 96
> 	enc cbc(aes) 0x963293b7f6b76f2dc512b316d9ae0886
> 	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

`ip xfrm policy` on right side MT-7621 device:

> src 192.168.2.0/24 dst 192.168.1.0/24 
> 	dir out priority 375423 
> 	tmpl src 10.10.10.2 dst 10.10.10.1
> 		proto esp spi 0xc245453b reqid 1 mode tunnel
> src 192.168.1.0/24 dst 192.168.2.0/24 
> 	dir fwd priority 375423 
> 	tmpl src 10.10.10.1 dst 10.10.10.2
> 		proto esp reqid 1 mode tunnel
> src 192.168.1.0/24 dst 192.168.2.0/24 
> 	dir in priority 375423 
> 	tmpl src 10.10.10.1 dst 10.10.10.2
> 		proto esp reqid 1 mode tunnel
> src 0.0.0.0/0 dst 0.0.0.0/0 
> 	socket in priority 0 
> src 0.0.0.0/0 dst 0.0.0.0/0 
> 	socket out priority 0 
> src 0.0.0.0/0 dst 0.0.0.0/0 
> 	socket in priority 0 
> src 0.0.0.0/0 dst 0.0.0.0/0 
> 	socket out priority 0 
> src ::/0 dst ::/0 
> 	socket in priority 0 
> src ::/0 dst ::/0 
> 	socket out priority 0 
> src ::/0 dst ::/0 
> 	socket in priority 0 
> src ::/0 dst ::/0 
> 	socket out priority 0 

_ipsec.conf_ file on right side MT-7621:

> config setup
>         strictcrlpolicy=no
>         uniqueids = no
>         charondebug="cfg 2, dmn 2, ike 2, net 2, lib 3, knl 4"
> 
> conn abc
>         authby = secret
>         type = tunnel
>         keyexchange = ikev2
>         ike = aes128-sha1-modp2048
>         auto = start
>         esp = aes128-sha1-modp2048
>         left = 10.10.10.2
>         leftsubnet = 192.168.2.0/24
>         right = 10.10.10.1
>         rightsubnet = 192.168.1.0/24
> 
> include /var/ipsec/ipsec.conf

_ipsec.conf_ file on left side MT-7621:

> config setup
>         strictcrlpolicy=no
>         uniqueids = no
> 
> conn 
>         authby = secret
>         type = tunnel
>         keyexchange = ikev2
>         ike = aes128-sha1-modp2048
>         auto = start
>         esp = aes128-sha1-modp2048
>         left = 10.10.10.1
>         leftsubnet = 192.168.1.0/24
>         right = 10.10.10.2
>         rightsubnet = 192.168.2.0/24

> include /var/ipsec/ipsec.conf

_/etc/config/firewall_ on right side MT-7621:

> config defaults
>         option input 'ACCEPT'
>         option output 'ACCEPT'
>         option forward 'ACCEPT'
>         option synflood_protect '1'
> 
> config zone
>         option name 'lan'
>         list network 'lan'
>         option input 'ACCEPT'
>         option output 'ACCEPT'
>         option forward 'ACCEPT'
> 
> config zone
>         option name 'wan'
>         list network 'wan'
>         list network 'wan6'
>         option output 'ACCEPT'
>         option masq '1'
>         option mtu_fix '1'
>         option input 'ACCEPT'
>         option forward 'ACCEPT'
> 
> config forwarding
>         option src 'lan'
>         option dest 'wan'
> 
> config rule
>         option name 'Allow-DHCP-Renew'
>         option src 'wan'
>         option proto 'udp'
>         option dest_port '68'
>         option target 'ACCEPT'
>         option family 'ipv4'
> 
> config rule
>         option name 'Allow-Ping'
>         option src 'wan'
>         option proto 'icmp'
>         option icmp_type 'echo-request'
>         option family 'ipv4'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-IGMP'
>         option src 'wan'
>         option proto 'igmp'
>         option family 'ipv4'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-DHCPv6'
>         option src 'wan'
>         option proto 'udp'
>         option dest_port '546'
>         option family 'ipv6'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-MLD'
>         option src 'wan'
>         option proto 'icmp'
>         option src_ip 'fe80::/10'
>         list icmp_type '130/0'
>         list icmp_type '131/0'
>         list icmp_type '132/0'
>         list icmp_type '143/0'
>         option family 'ipv6'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-ICMPv6-Input'
>         option src 'wan'
>         option proto 'icmp'
>         list icmp_type 'echo-request'
>         list icmp_type 'echo-reply'
>         list icmp_type 'destination-unreachable'
>         list icmp_type 'packet-too-big'
>         list icmp_type 'time-exceeded'
>         list icmp_type 'bad-header'
>         list icmp_type 'unknown-header-type'
>         list icmp_type 'router-solicitation'
>         list icmp_type 'neighbour-solicitation'
>         list icmp_type 'router-advertisement'
>         list icmp_type 'neighbour-advertisement'
>         option limit '1000/sec'
>         option family 'ipv6'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-ICMPv6-Forward'
>         option src 'wan'
>         option dest '*'
>         option proto 'icmp'
>         list icmp_type 'echo-request'
>         list icmp_type 'echo-reply'
>         list icmp_type 'destination-unreachable'
>         list icmp_type 'packet-too-big'
>         list icmp_type 'time-exceeded'
>         list icmp_type 'bad-header'
>         list icmp_type 'unknown-header-type'
>         option limit '1000/sec'
>         option family 'ipv6'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-IPSec-ESP'
>         option src 'wan'
>         option dest 'lan'
>         option proto 'esp'
>         option target 'ACCEPT'
> 
> config rule
>         option name 'Allow-ISAKMP'
>         option src 'wan'
>         option dest 'lan'
>         option dest_port '500'
>         option proto 'udp'
>         option target 'ACCEPT'

_/etc/config/firewall_ on left side MT-7621:

> config defaults
> 	option input 'ACCEPT'
> 	option output 'ACCEPT'
> 	option forward 'ACCEPT'
> 	option synflood_protect '1'
> 
> config zone
> 	option name 'lan'
> 	list network 'lan'
> 	option input 'ACCEPT'
> 	option output 'ACCEPT'
> 	option forward 'ACCEPT'
> 
> config zone
> 	option name 'wan'
> 	list network 'wan'
> 	list network 'wan6'
> 	option output 'ACCEPT'
> 	option masq '1'
> 	option mtu_fix '1'
> 	option input 'ACCEPT'
> 	option forward 'ACCEPT'
> 
> config forwarding
> 	option src 'lan'
> 	option dest 'wan'
> 
> config rule
> 	option name 'Allow-DHCP-Renew'
> 	option src 'wan'
> 	option proto 'udp'
> 	option dest_port '68'
> 	option target 'ACCEPT'
> 	option family 'ipv4'
> 
> config rule
> 	option name 'Allow-Ping'
> 	option src 'wan'
> 	option proto 'icmp'
> 	option icmp_type 'echo-request'
> 	option family 'ipv4'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-IGMP'
> 	option src 'wan'
> 	option proto 'igmp'
> 	option family 'ipv4'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-DHCPv6'
> 	option src 'wan'
> 	option proto 'udp'
> 	option dest_port '546'
> 	option family 'ipv6'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-MLD'
> 	option src 'wan'
> 	option proto 'icmp'
> 	option src_ip 'fe80::/10'
> 	list icmp_type '130/0'
> 	list icmp_type '131/0'
> 	list icmp_type '132/0'
> 	list icmp_type '143/0'
> 	option family 'ipv6'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-ICMPv6-Input'
> 	option src 'wan'
> 	option proto 'icmp'
> 	list icmp_type 'echo-request'
> 	list icmp_type 'echo-reply'
> 	list icmp_type 'destination-unreachable'
> 	list icmp_type 'packet-too-big'
> 	list icmp_type 'time-exceeded'
> 	list icmp_type 'bad-header'
> 	list icmp_type 'unknown-header-type'
> 	list icmp_type 'router-solicitation'
> 	list icmp_type 'neighbour-solicitation'
> 	list icmp_type 'router-advertisement'
> 	list icmp_type 'neighbour-advertisement'
> 	option limit '1000/sec'
> 	option family 'ipv6'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-ICMPv6-Forward'
> 	option src 'wan'
> 	option dest '*'
> 	option proto 'icmp'
> 	list icmp_type 'echo-request'
> 	list icmp_type 'echo-reply'
> 	list icmp_type 'destination-unreachable'
> 	list icmp_type 'packet-too-big'
> 	list icmp_type 'time-exceeded'
> 	list icmp_type 'bad-header'
> 	list icmp_type 'unknown-header-type'
> 	option limit '1000/sec'
> 	option family 'ipv6'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-IPSec-ESP'
> 	option src 'wan'
> 	option dest 'lan'
> 	option proto 'esp'
> 	option target 'ACCEPT'
> 
> config rule
> 	option name 'Allow-ISAKMP'
> 	option src 'wan'
> 	option dest 'lan'
> 	option dest_port '500'
> 	option proto 'udp'
> 	option target 'ACCEPT'

`iptables-save` on right-side MT-7621 device:

> *filter
> :INPUT ACCEPT [0:0]
> :FORWARD ACCEPT [0:0]
> :OUTPUT ACCEPT [0:0]
> -A INPUT -m policy --dir in --pol ipsec --proto esp -j ACCEPT
> -A FORWARD -s 192.168.2.0/24 -d 192.168.1.0/24 -m policy --dir out --pol ipsec --proto esp -j ACCEPT
> -A FORWARD -s 192.168.1.0/24 -d 192.168.2.0/24 -m policy --dir in --pol ipsec --proto esp -j ACCEPT
> -A OUTPUT -m policy --dir out --pol ipsec --proto esp -j ACCEPT
> COMMIT
> Completed on Tue Jan  3 22:15:38 2023
> Generated by iptables-save v1.8.7 on Tue Jan  3 22:15:38 2023
> *nat
> :PREROUTING ACCEPT [0:0]
> :INPUT ACCEPT [0:0]
> :OUTPUT ACCEPT [0:0]
> :POSTROUTING ACCEPT [0:0]
> COMMIT
> Completed on Tue Jan  3 22:15:38 2023
> Warning: iptables-legacy tables present, use iptables-legacy-save to see them

`iptables-save` on left-side MT-7621 device:

> Generated by iptables-save v1.8.7 on Tue Jan  3 22:10:26 2023
> *filter
> :INPUT ACCEPT [0:0]
> :FORWARD ACCEPT [0:0]
> :OUTPUT ACCEPT [0:0]
> -A INPUT -m policy --dir in --pol ipsec --proto esp -j ACCEPT
> -A FORWARD -s 192.168.1.0/24 -d 192.168.2.0/24 -m policy --dir out --pol ipsec --proto esp -j ACCEPT
> -A FORWARD -s 192.168.2.0/24 -d 192.168.1.0/24 -m policy --dir in --pol ipsec --proto esp -j ACCEPT
> -A OUTPUT -m policy --dir out --pol ipsec --proto esp -j ACCEPT
> COMMIT
> Completed on Tue Jan  3 22:10:26 2023
> Generated by iptables-save v1.8.7 on Tue Jan  3 22:10:26 2023
> *nat
> :PREROUTING ACCEPT [0:0]
> :INPUT ACCEPT [0:0]
> :OUTPUT ACCEPT [0:0]
> :POSTROUTING ACCEPT [0:0]
> -A POSTROUTING -s 192.168.1.0/24 -m policy --dir out --pol ipsec -j ACCEPT
> COMMIT
> Completed on Tue Jan  3 22:10:26 2023
> Warning: iptables-legacy tables present, use iptables-legacy-save to see them

Sorry for the long post, and thanks in advance!

**Comment by tobiasbrunner** [comment]
Please also post the output of `iptables-legacy-save`, there might be NAT rules that affect the traffic.

**Reply by umairsafdar768** [reply]
You are right, it was a problem with Masquerading on the 'wan' port. _fw4_ in openwrt masquerades all egress traffic by default
Solved it by excluding my traffic from this default behaviour using: 

> list masq_dest !10.10.10.0/24

in /etc/config/firewall of both gateways.
Thanks to this answer:
[https://forum.openwrt.org/t/ipsec-routing-firewall4/127101/6]([https://forum.openwrt.org/t/ipsec-routing-firewall4/127101/6](https://github.com/strongswan/strongswan/url))
